pkg/iam/iam.go

```diff
@@ -0,0 +1,116 @@
+		Region:      aws.String(policy.Region),
+		Credentials: credentials.NewSharedCredentials("", policy.Profile),
+	}
+
+	// Initialize session to AWS
+	sess, err := session.NewSession(awsConfig)
+	if err != nil {
+		return fmt.Errorf("session initialization for region %s and profile %s: %w", policy.Region, policy.Profile, err)
+	}
+	svc := iam.New(sess)
+
+	// Format the AWS Policy ARN
+	policyARN := fmt.Sprintf("arn:aws:iam::%s:policy/%s", policy.AccountID, policy.Name)
+
+	// Retrieve the Policy from AWS
+	receivedPolicy, err := svc.GetPolicy(&iam.GetPolicyInput{
+		PolicyArn: &policyARN,
+	})
+	if err != nil {
+		return fmt.Errorf("retrieve policy ARN %s: %w", policyARN, err)
+	}
+	log.Info(fmt.Sprintf("Received AWS Policy: %+v\n", receivedPolicy))
+
+	// Retrieve the current Policy version
+	currentVersion, err := svc.GetPolicyVersion(&iam.GetPolicyVersionInput{VersionId: receivedPolicy.Policy.DefaultVersionId, PolicyArn: aws.String(policyARN)})
+	if err != nil {
+		fmt.Println("Error", err)
+		return fmt.Errorf("retrieve policy version %s with policy ARN %s: %w", *receivedPolicy.Policy.DefaultVersionId, policyARN, err)
+	}
+	log.Info(fmt.Sprintf("Current Policy Version: %+v\n", currentVersion))
+
+	// URL decode to be able to Unmarshal into objects
+	jsonDocument, err := url.QueryUnescape(*currentVersion.PolicyVersion.Document)
+	if err != nil {
+		return fmt.Errorf("query unescape of: %s: %w", *currentVersion.PolicyVersion.Document, err)
+	}
+	document := PolicyDocument{}
+	json.Unmarshal([]byte(jsonDocument), &document)
+
+	// Append the new user to the Policy
+	document.appendStatement(userID, policy.AccountID, policy.Region)
+
+	// Marshal the updated policy document back to something AWS understands
+	jsonMarshal, err := json.Marshal(document)
+	if err != nil {
+		return fmt.Errorf("json marshal of: %s: %w", document, err)
+	}
+
+	// Create the new version of the Policy
+	true := true
+	_, err = svc.CreatePolicyVersion(&iam.CreatePolicyVersionInput{PolicyArn: aws.String(policyARN), PolicyDocument: aws.String(string(jsonMarshal)), SetAsDefault: &true})
+	if err != nil {
+		return fmt.Errorf("create policy version with arn %s: %w", policyARN, err)
+	}
+
+	// Delete the policy version to ensure that we don't succeed the maxium of 5 versions
+	_, err = svc.DeletePolicyVersion(&iam.DeletePolicyVersionInput{PolicyArn: aws.String(policyARN), VersionId: currentVersion.PolicyVersion.VersionId})
+	if err != nil {
+		return fmt.Errorf("delete policy version with arn %s: %w", policyARN, err)
+	}
+	return nil
+
+}
+
+func (p *PolicyDocument) appendStatement(userID, awsAccountID, region string) {
+	s := StatementEntry{
+		Effect:    "Allow",
+		Action:    []string{"rds-db:connect"},
+		Resource:  []string{fmt.Sprintf("arn:aws:rds-db:%s:%s:dbuser:*/iam_developer_%s", region, awsAccountID, userID)},
+		Condition: StringLike{StringLike: UserID{AWSUserID: fmt.Sprintf("*:%s@lunarway.com", userID)}},
+	}
+	p.Statement = append(p.Statement, s)
```

Comment:
We need to be more clever about this. Otherwise changes to a CR will always add a new statement with the user.
In other words we should only add the statement if it does not already exist.

Reply:
And tests of this would be nice :D